### PR TITLE
neard: recompress_storage: clean out unnecessary columns

### PR DESCRIPTION
**Pull request by mina86**
When recompressing database of an archival node, skip
ColPartialChunks, ColInvalidChunks and ColTrieChanges columns which
can be safely deleted.  Data in the first one can be reconstructed
from ColChunks, ColInvalidChunks is only needed at head and the last
is never read by archival nodes.

Mostly for testing, if someone wants to keep those columns,
offer --keep-partial-chunks, --keep-invalid-chunks
and --keep-trie-changes switches.  They are always on when dealing
with non-archival node.

Issue: https://github.com/near/nearcore/issues/6119
Issue: https://github.com/near/nearcore/issues/6242
Issue: https://github.com/near/nearcore/issues/6250

### CODE REVIEW

File: neard/src/cli.rs [hunk 1]
```diff
@@ -493,12 +493,28 @@ pub(super) struct RecompressStorageSubCommand {
...
+
+    /// Delete data from ColPartialChunks column.  The option is ignored unless
+    /// running on archival node.
+    #[clap(long)]
+    clean_partial_chunks: bool,
+
+    /// Clean ColTrieChanges column.  The option is ignored unless running on
+    /// archival node.
+    #[clap(long)]
+    clean_trie_changes: bool,
```
Comment (bowenwang1996) [inline neard/src/cli.rs:505]: I think it might be easier if they default to true. What do you think?

File: nearcore/src/lib.rs [hunk 2]
```diff
@@ -500,9 +500,27 @@ pub fn start_with_config_and_synchronization(
...
+        .map_err(|err| anyhow::anyhow!("{}: {}", config_path.display(), err))?
+        .archive;
+    let mut skip_columns = Vec::new();
+    if archive && opts.clean_partial_chunks {
+        skip_columns.push(near_store::db::DBCol::ColPartialChunks);
+    }
+    if archive && opts.clean_trie_changes {
+        skip_columns.push(near_store::db::DBCol::ColTrieChanges);
+    }
+
```
Comment (bowenwang1996) [inline nearcore/src/lib.rs:523]: Can we print a warning here that this may take a day to finish?

File: neard/src/cli.rs [hunk 3]
```diff
@@ -493,12 +493,29 @@ pub(super) struct RecompressStorageSubCommand {
...
+    /// Keep data in ColTrieChanges column.  This is always true if node is
+    /// not an archival node.
+    #[clap(long)]
+    keep_trie_changes: bool,
 }
 
 impl RecompressStorageSubCommand {
     pub(super) fn run(self, home_dir: &Path) {
-        if let Err(err) = nearcore::recompress_storage(&home_dir, &self.output_dir) {
+        warn!(target: "neard", "Recompressing storage; note that this operation may take up to a day to finish.");
```
Comment (mm-near) [inline neard/src/cli.rs:510]: and if it is disturbed, you have to start from scratch, right?

File: neard/src/cli.rs [hunk 4]
```diff
@@ -493,12 +493,29 @@ pub(super) struct RecompressStorageSubCommand {
     /// Directory where to save new storage.
     #[clap(long)]
     output_dir: PathBuf,
+
+    /// Keep data in ColPartialChunks column.  This is always true if node is
+    /// not an archival node.
+    #[clap(long)]
+    keep_partial_chunks: bool,
```
Comment (mm-near) [inline neard/src/cli.rs:500]: What is the default value here? false?

so if someone runs the recompress after we submit this PR (assuming they have archival node) - they will automatically 'lose' those columns, right?

Maybe we should wait with this PR until we finish the code for computing partial_chunks/trie changes on the fly?

File: neard/src/cli.rs [hunk 5]
```diff
@@ -493,12 +493,29 @@ pub(super) struct RecompressStorageSubCommand {
...
+
+    /// Keep data in ColTrieChanges column.  This is always true if node is
+    /// not an archival node.
+    #[clap(long)]
+    keep_trie_changes: bool,
```
Comment (mm-near) [inline neard/src/cli.rs:505]: maybe we should put 'archival' somewhere in the name?

**Review by mm-near — CHANGES_REQUESTED** [review]

File: neard/src/cli.rs (hunk 3 above)
Comment (mina86) [reply neard/src/cli.rs:510]: Yep.

File: neard/src/cli.rs (hunk 4 above)
Comment (mina86) [reply neard/src/cli.rs:500]: > What is the default value here? false?

Yep.

> so if someone runs the recompress after we submit this PR (assuming they have archival node) - they will automatically 'lose' those columns, right?

Yep.

> Maybe we should wait with this PR until we finish the code for computing partial_chunks/trie changes on the fly?

The code to recompute partial chunks is already merged in #6377. For recomputing trie changes we decided we won’t bother at least for now. The column only matters when non-archival node is doing GC.

There is #6439 still pending but I think it doesn’t matter that much which one of those gets merged first.

File: neard/src/cli.rs (hunk 5 above)
Comment (mina86) [reply neard/src/cli.rs:505]: Do you have idea for the names? I think the current one is descriptive enough.

File: nearcore/src/lib.rs [hunk 6]
```diff
@@ -589,9 +624,27 @@ pub fn recompress_storage(home_dir: &Path, dst_dir: &Path) -> anyhow::Result<()>
...
+        let tip: Option<near_primitives::block::Tip> =
+            src_store.get_ser(DBCol::ColBlockMisc, near_store::FINAL_HEAD_KEY)?;
+        if let Some(tip) = tip {
+            let chunk_tail: near_primitives::types::BlockHeight = tip.height;
+            info!("Setting chunk tail to {}", chunk_tail);
+            let mut store_update = dst_store.store_update();
+            store_update.set_ser(DBCol::ColBlockMisc, near_store::CHUNK_TAIL_KEY, &chunk_tail)?;
+            store_update.commit()?;
+        } else {
+            info!("Final head not set; leaving chunk tail be");
```
Comment (mm-near) [inline nearcore/src/lib.rs:641]: what are the scenarios when we're missing the 'final head'?

File: neard/src/cli.rs [hunk 7]
```diff
@@ -465,12 +465,37 @@ pub(super) struct RecompressStorageSubCommand {
...
 
 impl RecompressStorageSubCommand {
     pub(super) fn run(self, home_dir: &Path) {
-        if let Err(err) = nearcore::recompress_storage(&home_dir, &self.output_dir) {
+        warn!(target: "neard", "Recompressing storage; note that this operation may take up to a day to finish.");
+        let opts = nearcore::RecompressOpts {
+            dest_dir: self.output_dir,
+            keep_partial_chunks: self.keep_partial_chunks,
+            keep_invalid_chunks: self.keep_invalid_chunks,
+            keep_trie_changes: self.keep_trie_changes,
```
Comment (mm-near) [inline neard/src/cli.rs:494]: for this column - let's figure out the cold storage migration first.

**Review by mm-near — DISMISSED** [review]

File: nearcore/src/lib.rs (hunk 6 above)
Comment (mina86) [reply nearcore/src/lib.rs:641]: In any legitimate uses of recompress storage I can’t really think of any scenario. Do you think the code should just panic?

File: neard/src/cli.rs (hunk 7 above)
Comment (mina86) [reply neard/src/cli.rs:495]: After discussion we’ve agreed this column can go.

File: nearcore/src/lib.rs [hunk 8]
```diff
@@ -589,9 +639,21 @@ pub fn recompress_storage(home_dir: &Path, dst_dir: &Path) -> anyhow::Result<()>
         store_update.commit()?;
     }
 
+    // If we’re not keeping ColPartialChunks, update chunk tail to point to
+    // current final block.  If we don’t do that, the gc will try to work its
+    // way from the genesis even though chunks at those heights have been
+    // deleted.
+    if skip_columns.contains(&DBCol::ColPartialChunks) {
+        let chunk_tail = final_head_height.unwrap();
+        info!("Setting chunk tail to {}", chunk_tail);
```
Comment (nagisa) [inline nearcore/src/lib.rs:648]:
```suggestion
        info!(target: "recompress"; %chunk_tail, "Setting chunk tail");
```

File: nearcore/src/lib.rs [hunk 9]
```diff
@@ -537,21 +561,47 @@ pub fn recompress_storage(home_dir: &Path, dst_dir: &Path) -> anyhow::Result<()>
...
 
     for (n, column) in DBCol::iter().enumerate() {
+        if skip_columns.contains(&column) {
+            info!(
+                "Clearing out col{} ‘{}’ ({:2} / {:2})",
+                column as usize,
+                column,
+                n + 1,
+                DBCol::COUNT
+            );
```
Comment (nagisa) [inline nearcore/src/lib.rs:601]:
```suggestion
            info!(
                column_id = column as usize,
                %column,
                current_column = n + 1,
                column_count = DBCol::COUNT,
                "Clearing a column",                
            );
```

**Pull request merged**